plugins/quickbooks-online/src/src/Facade/QuickBooksFacade.php

```diff
@@ -376,143 +357,486 @@ function (array $a, array $b) {
-                                'Payment ID: %s export failed with error %s.',
-                                $ucrmPayment['id'],
+                                'Payment to apply credit export failed with error %s',
                                 $exception->getMessage()
                             )
                         );
                     }
                 }
             }
 
-            // update last processed payment in the end
-            $pluginData->lastExportedPaymentID = $ucrmPayment['id'];
+            $pluginData->lastExportedCreditID = $ucrmCredit['id'];
             $this->optionsManager->update();
+        } // end foreach $ucrmCredit
+    }
+
+    private function getPaymentCovers($payments, DataService $dataService): array {
+        $lineArray = null;
+        $additionalUnapplied = 0.0;
+        $totalApplied = 0.0;
+        foreach ($payments as $paymentCovers) {
+            if ($paymentCovers['amount'] == 0) continue;
+
+            if ($paymentCovers['refundId']) {
+                $this->logger->notice('Payment has refundId, not yet supported! Will set as unapplied');
+                $additionalUnapplied += $paymentCovers['amount'];
+                continue;
+            }
+            if ($paymentCovers['invoiceId'] == '') {
+                $this->logger->notice('Payment has empty invoiceId! Will set as unapplied');
+                $additionalUnapplied += $paymentCovers['amount'];
+                continue;
+            }
+
+            $invId = $paymentCovers['invoiceId'];
+            $invoices = $this->dataServiceQuery($dataService,"SELECT * FROM INVOICE WHERE DOCNUMBER LIKE '%/$invId'");
+            if (!$invoices)
+                $invoices = $this->dataServiceQuery($dataService,"SELECT * FROM INVOICE WHERE DOCNUMBER = '$invId'");
+
+            if (!$invoices) {
+                $this->logger->warning(sprintf('Unable to find invoiceId %s covered by payment, will set as unapplied', $paymentCovers['invoiceId']));
+                $additionalUnapplied += $paymentCovers['amount'];
+                continue;
+            }
+
+            $lineArray[] = Line::create(
+                [
+                    'Amount' => $paymentCovers['amount'],
+                    'LinkedTxn' => [
+                        'TxnId' => $invoices[0]->Id,
+                        'TxnType' => 'Invoice',
+                    ],
+                ]
+            );
+
+            $totalApplied += $paymentCovers['amount'];
+            $this->logger->debug("Payment applying \${$paymentCovers['amount']} to Invoice {$invoices[0]->DocNumber}");
+        }
+
+        return [$lineArray, $additionalUnapplied, $totalApplied];
+    }
+
+    private $itemCache = [];
+    private function getItems($items, int $qbIncomeAccountId, bool $negateQty, DataService $dataService, PluginData $pluginData): array {
+        $lines = [];
+        foreach ($items as $item) {
+            $qbItem = $this->itemCache[$item['label']];
```

Comment:
this will throw an undefined index error if the item doesn't exist in cache, you can use null coalesce operator to handle it
```suggestion
            $qbItem = $this->itemCache[$item['label']] ?? false;
```

Reply:
I'm using the code without issue. I thought what I found with PHP (didn't have experience in PHP before) was that checking `if (! $qbItem)` like I did in L668 returns false if it's null or if it's false. I thought this type of check was used constantly for null throughout the file.

But I could still change it if you want me to

Reply:
Sorry, not an error, just a notice, I'm too used to our base code converting even notices to errors. :) However even if it works, it's still bad practice.

You can check this example code over at https://3v4l.org/I0uOJ
```php
<?php

$itemCache = [];

$qbItem = $itemCache['foo'];
if (! $qbItem) {
    var_dump('foo works, but shows notice');
}

$qbItem = $itemCache['bar'] ?? false;
if (! $qbItem) {
    var_dump('bar works and no notice');
}

$qbItem = array_key_exists('foobar', $itemCache) ? $itemCache['foobar'] : false;
if (! $qbItem) {
    var_dump('foobar works and no notice');
}

$qbItem = isset($itemCache['foo2']) ? $itemCache['foo2'] : false;
if (! $qbItem) {
    var_dump('foo2 works and no notice, but hides falsey values too, see foo3');
}

$itemCache = ['foo3' => 0];
$qbItem = isset($itemCache['foo3']) ? $itemCache['foo3'] : false;
if (! $qbItem) {
    var_dump('foo3 has no notice, but hides falsey values');
}
```

Reply:
Done